core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -119,6 +119,8 @@ class ReplicaFetcherThread(name: String,
 
       if (fetchOffset != replica.logEndOffset.messageOffset)
         throw new RuntimeException("Offset mismatch for partition %s: fetched offset = %d, log end offset = %d.".format(topicAndPartition, fetchOffset, replica.logEndOffset.messageOffset))
+      if (!messageSet.isEmpty && messageSet.head.offset < replica.logEndOffset.messageOffset)
+        throw new RuntimeException("Offset error for partition %s: first offset in message set = %d, must not be less than log end offset = %d.".format(topicAndPartition, messageSet.head.offset, replica.logEndOffset.messageOffset))
```

Comment:
@imandhan : Yes, you are right. I didn't realize that we already had a check there and a unit test for it. Sorry for the confusion. I will close the jira and mark it as a non-issue. Hopefully that you can find some other Kafka jiras to contribute.